include/net/http.h

```diff
@@ -9,45 +9,388 @@
+ * @param user_data A valid pointer on some user data or NULL
+ */
+typedef void (*http_request_cb_t)(struct http_client_ctx *ctx,
+				  u8_t *data, size_t buflen,
+				  size_t datalen,
+				  enum http_final_call final_data,
+				  void *user_data);
+
+/**
+ * HTTP client context information. This contains all the data that is
+ * needed when doing HTTP requests.
+ */
+struct http_client_ctx {
+	struct http_parser parser;
+	struct http_parser_settings settings;
+
+	/** Server name */
+	const char *server;
+
+#if defined(CONFIG_DNS_RESOLVER)
+	/** Remember the DNS query id so that it can be cancelled
+	 * if the HTTP context is released and the query is active
+	 * at that time.
+	 */
+	u16_t dns_id;
+#endif
+
+	struct {
+		/** Local socket address */
+		struct sockaddr local;
+
+		/** Remote (server) socket address */
+		struct sockaddr remote;
+
+		/** IP stack network context */
+		struct net_context *ctx;
+
+		/** Network timeout */
+		s32_t timeout;
+
+		/** User can define this callback if it wants to have
+		 * special handling of the received raw data.
+		 */
+		http_receive_cb_t receive_cb;
+	} tcp;
+
+	/** HTTP request information */
+	struct {
+		/**
+		 * Semaphore to signal HTTP request completion
+		 */
+		struct k_sem wait;
+
+		/** User provided HTTP request callback */
+		http_request_cb_t cb;
+
+		/** Hostname to be used in the request */
+		const char *host;
+
+		/** User provided data */
+		void *user_data;
+
+		/** What method we used here (GET, POST, HEAD etc.)
+		 */
+		enum http_method method;
+	} req;
+
+	/** HTTP response information */
+	struct {
+		/** Where the result is stored, this is to be provided
+		 * by the user.
+		 */
+		u8_t *result_buf;
+
+		/** Where the body starts.
+		 */
+		u8_t *body_start;
+
+		/** Result buffer maximum length */
+		size_t result_buf_len;
+
+		/** Length of the data in the result buf. If the value is
+		 * larger than result_buf_len, then it means that the data
+		 * is truncated and could not be fully copied into result_buf
+		 */
+		size_t data_len;
+
+		/** HTTP Content-Length field value */
+		size_t content_length;
+
+		/** Content length parsed. This should be the same as the
+		 * content_length field if parsing was ok.
+		 */
+		size_t processed;
+
+		/* https://tools.ietf.org/html/rfc7230#section-3.1.2
+		 * The status-code element is a 3-digit integer code
+		 *
+		 * The reason-phrase element exists for the sole purpose of
+		 * providing a textual description associated with the
+		 * numeric status code. A client SHOULD ignore the
+		 * reason-phrase content.
+		 */
+		char http_status[HTTP_STATUS_STR_SIZE];
+
+		u8_t cl_present:1;
+		u8_t body_found:1;
+	} rsp;
+};
+
+/**
+ * HTTP client request. This contains all the data that is needed when doing
+ * a HTTP request.
+ */
 struct http_client_request {
-	/** The HTTP method: GET, HEAD, OPTIONS, POST */
-	char *method;
+	/** The HTTP method: GET, HEAD, OPTIONS, POST, ... */
+	const char *method;
```

Comment:
Leftover from HTTP client application, it wanted to add a space after the method which required a separate string. I will investigate if we can use directly the strings that are available from the http_parser.c.